cartographer/mapping/internal/3d/local_trajectory_builder_3d.cc

```diff
@@ -134,111 +133,123 @@ LocalTrajectoryBuilder3D::AddRangeData(
-
-    const auto voxel_filter_start = std::chrono::steady_clock::now();
-    const sensor::RangeData filtered_range_data = {
-        current_pose.translation(),
-        sensor::VoxelFilter(options_.voxel_filter_size())
-            .Filter(accumulated_range_data_.returns),
-        sensor::VoxelFilter(options_.voxel_filter_size())
-            .Filter(accumulated_range_data_.misses)};
-    const auto voxel_filter_stop = std::chrono::steady_clock::now();
-    const auto voxel_filter_duration = voxel_filter_stop - voxel_filter_start;
+  CHECK(std::next(hits_poses_it) == hits_poses.end());
 
-    if (sensor_duration.has_value()) {
-      const double voxel_filter_fraction =
-          common::ToSeconds(voxel_filter_duration) /
-          common::ToSeconds(sensor_duration.value());
-      kLocalSlamVoxelFilterFraction->Set(voxel_filter_fraction);
-    }
+  const common::Time current_sensor_time = synchronized_data.time;
+  absl::optional<common::Duration> sensor_duration;
+  if (last_sensor_time_.has_value()) {
+    sensor_duration = current_sensor_time - last_sensor_time_.value();
+  }
+  last_sensor_time_ = current_sensor_time;
+
+  const common::Time current_time = hit_times.back();
+  const auto voxel_filter_start = std::chrono::steady_clock::now();
+  const sensor::RangeData filtered_range_data = {
+      extrapolation_result.current_pose.translation().cast<float>(),
+      sensor::VoxelFilter(options_.voxel_filter_size())
+          .Filter(accumulated_range_data.returns),
+      sensor::VoxelFilter(options_.voxel_filter_size())
+          .Filter(accumulated_range_data.misses)};
+  const auto voxel_filter_stop = std::chrono::steady_clock::now();
+  const auto voxel_filter_duration = voxel_filter_stop - voxel_filter_start;
 
-    return AddAccumulatedRangeData(
-        current_sensor_time,
-        sensor::TransformRangeData(filtered_range_data, current_pose.inverse()),
-        sensor_duration);
+  if (sensor_duration.has_value()) {
+    const double voxel_filter_fraction =
+        common::ToSeconds(voxel_filter_duration) /
+        common::ToSeconds(sensor_duration.value());
+    kLocalSlamVoxelFilterFraction->Set(voxel_filter_fraction);
   }
-  return nullptr;
+
+  return AddAccumulatedRangeData(
+      current_time,
+      sensor::TransformRangeData(
+          filtered_range_data,
+          extrapolation_result.current_pose.inverse().cast<float>()),
+      sensor_duration, extrapolation_result.current_pose,
+      extrapolation_result.current_velocity,
+      extrapolation_result.gravity_from_tracking);
 }
 
 std::unique_ptr<LocalTrajectoryBuilder3D::MatchingResult>
 LocalTrajectoryBuilder3D::AddAccumulatedRangeData(
     const common::Time time,
     const sensor::RangeData& filtered_range_data_in_tracking,
-    const absl::optional<common::Duration>& sensor_duration) {
+    const absl::optional<common::Duration>& sensor_duration,
+    const transform::Rigid3d& pose_prediction,
+    const Eigen::Vector3d& current_velocity,
```

Comment:
Why is `current_velocity` given as an argument? It seems to be unused.

Reply:
Good catch, that was unrelated. Removed it.